include/engine/routing_algorithms/routing_init.hpp

```diff
@@ -0,0 +1,334 @@
+#ifndef OSRM_ENGINE_ROUTING_INIT_HPP
+#define OSRM_ENGINE_ROUTING_INIT_HPP
+
+#include "engine/routing_algorithms/routing_base_ch.hpp"
+#include "engine/routing_algorithms/routing_base_mld.hpp"
+
+namespace osrm
+{
+namespace engine
+{
+namespace routing_algorithms
+{
+
+namespace detail
+{
+template <typename Facade, typename Heap>
+void relaxSourceEdges(const Facade &facade, Heap &heap, const NodeID node, const EdgeWeight weight)
+{
+    for (auto edge : facade.GetAdjacentEdgeRange(node))
+    {
+        const auto &edge_data = facade.GetEdgeData(edge);
+        if (edge_data.forward)
+        {
+            const auto to = facade.GetTarget(edge);
+            const auto to_weight = weight + getEdgeInternalWeight(facade, node, edge);
+            if (!heap.WasInserted(to))
+            {
+                heap.Insert(to, to_weight, node);
+            }
+            else if (!heap.WasRemoved(to) && to_weight < heap.GetKey(to))
```

Comment:
Here node `to` can be removed if it is a source node (inserted via `InsertedVisited`) and has not a weight, so `heap.GetKey(to)` returns a weight from the projected point to the end of node.

For example, for an ebg node `a------->b` let the snapped node is `a`, forward node is `0` and reverse node is `1`. Let `weight(ab)` is 100, so node `0` will be inserted with weight 100, the next  node `1` will be inserted with weight 0. If there is a forward u-turn edge from `1` to `0` with weight smaller than `weight(ab)` then  `to_weight < heap.GetKey(to)` is true and weight 100 will be changed to some wrong one.